### ISSUE TITLE
ComplexityValidationRule deadlock

### ISSUE THREAD

**Issue by arguit** [graphql-dotnet/graphql-dotnet#3527]
## Description

After adding `ComplexityValidationRule` to the `ValidationRules` as suggested in the migration (v7) documentation process of validation of query with `RecursiveFragment` ends in an infinite loop in `ComplexityAnalyzer`'s method `GetDependencies`.

My understanding of the issue:

In v5 the ComplexityValidation was done after ValidationRules so the `NoFragmentCycles` validation rule ended the validation process with error before complexity analyzer.

In v7 the ComplexityValidation was transformed into ValidationRule and it runs with all the other core / default validation rules asynchronously. So even though the `NoFragmentCycles` detects the cycle correctly the ComplexityValidationRule runs simultaneously and never ends.

### Steps to reproduce

1. Add `ComplexityValidationRule` to `ValidationRules` of `ExecutionOptions`
2. Execute query like this
```
                   {
                        type_All(limit: 20) { 
                            items { 
                                links(limit: 20) { 
                                    items { 
                                        ... RecursiveFragment
                                    }
                                }
                            } 
                        }
                    } 
                    fragment RecursiveFragment on Type   {
                       links(limit: 20) {
                            items {
                                ... RecursiveFragment
                            }
                        }
                    }
```

### Expected result

It should end with `Cannot spread fragment 'RecursiveFragment' within itself.` and end the process of validation somehow.

### Actual result

Deadlock in of the following method.

1. There is one `_selectionSetsToVisit`
2. It `Pop()` out
3. Then it `Push()` another one back in
4. Goes back to 1.

```
HashSet<GraphQLFragmentDefinition>? GetDependencies(GraphQLFragmentDefinition def)
            {
                HashSet<GraphQLFragmentDefinition>? dependencies = null;
                _selectionSetsToVisit.Push(def.SelectionSet);

                while (_selectionSetsToVisit.Count > 0)
                {
                    foreach (var selection in _selectionSetsToVisit.Pop().Selections)
                    {
                        if (selection is GraphQLFragmentSpread spread)
                        {
                            var frag = document.FindFragmentDefinition(spread.FragmentName.Name.Value);
                            if (frag != null)
                            {
                                (dependencies ??= new()).Add(frag);
                                _selectionSetsToVisit.Push(frag.SelectionSet);
                            }
                        }
                        else if (selection is IHasSelectionSetNode hasSet && hasSet.SelectionSet != null)
                        {
                            _selectionSetsToVisit.Push(hasSet.SelectionSet);
                        }
                    }
                }

                _selectionSetsToVisit.Clear();
                return dependencies;
            }
```

### Environment

Win10

**Comment by sungam3r** [comment]
@Shane32 It looks like one more iteration fixing complexity analyzer. @arguit Could you please post a PR? It would be enough to just provide a test that verifies aforementioned issue.

See
- https://github.com/graphql-dotnet/graphql-dotnet/pull/3222
- https://github.com/graphql-dotnet/graphql-dotnet/pull/3209
- https://github.com/graphql-dotnet/graphql-dotnet/issues/3178

I do not intend to fix this myself, but I can help with review. Now I'm deep into OTel.

**Comment by arguit** [comment]
@sungam3r Basic test to demonstrate the issue added.

**Comment by sungam3r** [comment]
@arguit You may try preview package from https://github.com/orgs/graphql-dotnet/packages?repo_name=graphql-dotnet
It think that 7.3 release will be published ASAP after merging #3467.

**Comment by sungam3r** [comment]
This one https://github.com/graphql-dotnet/graphql-dotnet/pkgs/nuget/GraphQL/70002870
